File: src/Butil/Bit.Butil/Publics/Hid.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/WebHID_API">WebHID API</see>
/// (<c>navigator.hid</c>): talk to a human-interface device the operating system's own drivers do
/// not fully expose - a macro pad, a flight yoke, a vendor-defined control surface.
/// </summary>
/// <remarks>
/// Chromium only, and only over HTTPS. <see cref="RequestDevice"/> opens the browser's device
/// chooser and so must run inside a user gesture.
/// <br/>
/// The browser blocks devices whose usage pages would let a page impersonate the user's own input -
/// keyboards and mice among them - so WebHID is for the long tail, not for reading the keyboard.
/// </remarks>
[ButilService(typeof(Hid))]
public class Hid(IJSRuntime js) : IAsyncDisposable
{
    internal const string InputReportMethodName = nameof(InvokeHidInputReport);
    internal const string ConnectedMethodName = nameof(InvokeHidConnected);
    internal const string DisconnectedMethodName = nameof(InvokeHidDisconnected);

    // Keyed by subscription id, but carrying the device each one belongs to: hid.ts release() detaches
    // that device's listeners, so the .NET closures behind them have to go at the same moment -
    // otherwise disposing one device leaks its handlers for as long as the service lives.
    private readonly ConcurrentDictionary<Guid, (string DeviceId, Action<HidInputReport> Handler)> _inputHandlers = new();
    private readonly ConcurrentDictionary<Guid, (Action<HidDevice>? OnConnected, Action<HidDevice>? OnDisconnected)> _connectionHandlers = new();

    // Every handle this service handed out, so a scope/circuit teardown closes the device even
    // when the caller never disposed it.
    private readonly ConcurrentDictionary<string, HidDevice> _devices = new();

    // Per-instance callback reference (see Keyboard): listeners are isolated per circuit / WASM app
    // and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<Hid>? _dotNetRef;
    private DotNetObjectReference<Hid> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>navigator.hid</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.hid.isSupported");

    /// <summary>
    /// Opens the browser's device chooser and returns the devices the user picked - an array,
    /// because one physical device can present as several. Empty when they dismissed it. Must be
    /// called from a user gesture.
    /// </summary>
    /// <param name="filters">Which devices the chooser lists. Pass none to list everything the browser allows.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidDeviceInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidCollectionInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidReportInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidDeviceFilter))]
    public async ValueTask<HidDevice[]> RequestDevice(params HidDeviceFilter[] filters)
    {
        var infos = await js.Invoke<HidDeviceInfo[]>("BitButil.hid.requestDevice", (object)filters);
        return [.. infos.Select(Track)];
    }

    /// <summary>The devices this origin has already been granted, without showing a chooser.</summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidDeviceInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidCollectionInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidReportInfo))]
    public async ValueTask<HidDevice[]> GetDevices()
    {
        var infos = await js.Invoke<HidDeviceInfo[]>("BitButil.hid.getDevices");
        return [.. infos.Select(Track)];
    }

    // One handle per browser-side id. hid.ts idOf() deliberately hands the same id back for the same
    // device, so a device surfaced again by GetDevices or by a connect event has to return the handle
    // already in play: a second wrapper over one JS registry entry would let disposing either of them
    // release the device out from under the other. Info stays the snapshot the first handle was
    // created from - it is documented as not updating on its own, and GetInfo() re-reads it.
    private HidDevice Track(HidDeviceInfo info) => _devices.GetOrAdd(info.Id, _ => new HidDevice(this, js, info));

    // Called by a handle that is disposing itself, so the service stops holding it. The JS-side
    // release() detaches every input listener attached through that handle, so the .NET closures
    // behind them have to go with it - otherwise disposing one device leaks the handlers of that
    // device for as long as the service lives. The registry removal is identity-checked so a handle
    // that has already been replaced cannot untrack its successor.
    internal void Forget(HidDevice device)
    {
        _devices.TryRemove(new KeyValuePair<string, HidDevice>(device.Id, device));

        foreach (var pair in _inputHandlers)
        {
            if (pair.Value.DeviceId == device.Id) _inputHandlers.TryRemove(pair.Key, out _);
        }
    }

    /// <summary>
    /// Invoked from JS for each input report. Public + <see cref="JSInvokableAttribute"/> so it can
    /// be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(InputReportMethodName)]
    public void InvokeHidInputReport(Guid id, HidInputReport report)
    {
        if (_inputHandlers.TryGetValue(id, out var entry)) entry.Handler.Invoke(report);
    }

    /// <summary>Invoked from JS on <c>navigator.hid</c>'s <c>connect</c>. See <see cref="InvokeHidInputReport"/>.</summary>
    [JSInvokable(ConnectedMethodName)]
    public void InvokeHidConnected(Guid id, HidDeviceInfo info)
    {
        if (_connectionHandlers.TryGetValue(id, out var handlers)) handlers.OnConnected?.Invoke(Track(info));
    }

    /// <summary>Invoked from JS on <c>disconnect</c>. See <see cref="InvokeHidInputReport"/>.</summary>
    [JSInvokable(DisconnectedMethodName)]
    public void InvokeHidDisconnected(Guid id, HidDeviceInfo info)
    {
        if (_connectionHandlers.TryGetValue(id, out var handlers)) handlers.OnDisconnected?.Invoke(Track(info));
    }

    [DynamicDependency(nameof(InvokeHidInputReport), typeof(Hid))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidInputReport))]
    internal async ValueTask<ButilSubscription> SubscribeInputReports(string deviceId, Action<HidInputReport> handler)
    {
        return await ButilSubscriptionHelper.Register(_inputHandlers, (deviceId, handler),
                                                      id => js.InvokeRegister("BitButil.hid.subscribeInputReports", id, DotNetRef, deviceId),
                                                      id => js.InvokeVoid("BitButil.hid.unsubscribeInputReports", id),
                                                      "The input report listener could not be attached - the device handle is no longer known.");
    }

    /// <summary>
    /// Watches devices being plugged in and unplugged. Only devices this origin already has
    /// permission for raise these.
    /// </summary>
    /// <returns>A subscription - dispose it to detach the listeners.</returns>
    /// <exception cref="InvalidOperationException">The listeners were not attached - no WebHID support.</exception>
    [DynamicDependency(nameof(InvokeHidConnected), typeof(Hid))]
    [DynamicDependency(nameof(InvokeHidDisconnected), typeof(Hid))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidDeviceInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidCollectionInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(HidReportInfo))]
    public async ValueTask<ButilSubscription> SubscribeConnection(Action<HidDevice>? onConnected = null,
                                                                  Action<HidDevice>? onDisconnected = null)
    {
        return await ButilSubscriptionHelper.Register(_connectionHandlers, (onConnected, onDisconnected),
                                                      id => js.InvokeRegister("BitButil.hid.subscribeConnection", id, DotNetRef),
                                                      id => js.InvokeVoid("BitButil.hid.unsubscribeConnection", id),
                                                      "The HID connection listeners could not be attached - the WebHID API is not available.");
    }

    /// <summary>
    /// Closes every device this service handed out, detaches its listeners and releases the
    /// interop reference.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            // Releasing a device detaches its input listeners, and every entry left here belongs to a
            // device still in _devices - Forget() strips the rest - so the loop below covers them.
            _inputHandlers.Clear();

            foreach (var id in _connectionHandlers.Keys.ToArray())
            {
                _connectionHandlers.TryRemove(id, out _);
                await js.InvokeVoid("BitButil.hid.unsubscribeConnection", id);
            }

            foreach (var device in _devices.Values.ToArray())
            {
                await device.DisposeAsync();
            }
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }

        GC.SuppressFinalize(this);
    }
}

File: src/Butil/Bit.Butil/Scripts/hid.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // HIDDevice objects carry the user's grant and stay on this side of the boundary; .NET works
    // through the Guid it minted for each one. See bluetooth.ts for the same shape.
    const _devices: { [id: string]: any } = {};
    const _inputListeners: { [id: string]: { device: any, handler: EventListener } } = {};
    const _connectionListeners: { [id: string]: { connect: EventListener, disconnect: EventListener } } = {};

    // Handle ids are minted here rather than by .NET, through the registry every device module
    // shares: the browser surfaces a device .NET has never seen, so there is nothing for it to key on
    // until the info comes back. Every entry point that hands one out hands out its id with it.
    const _registry = butil.utils.handleRegistry('hid', _devices);
    const idOf = _registry.idOf;


    butil.hid = {
        isSupported() { return !!(navigator as any).hid; },
        requestDevice,
        getDevices,
        forget,
        release,
        open,
        close,
        isOpened,
        getInfo,
        sendReport,
        sendFeatureReport,
        receiveFeatureReport,
        subscribeInputReports,
        unsubscribeInputReports,
        subscribeConnection,
        unsubscribeConnection
    };

    function hid() { return (navigator as any).hid; }

    function reports(list: any[]) {
        return (list ?? []).map((report: any) => ({
            reportId: report.reportId,
            // Only the item count is carried across: a full HID item descriptor is a deep tree
            // that nothing on the .NET side can act on without the raw report bytes anyway.
            itemCount: (report.items ?? []).length
        }));
    }

    function describe(id: string, device: any) {
        return {
            id,
            vendorId: device.vendorId,
            productId: device.productId,
            productName: device.productName ?? null,
            opened: !!device.opened,
            collections: (device.collections ?? []).map((collection: any) => ({
                usagePage: collection.usagePage,
                usage: collection.usage,
                inputReports: reports(collection.inputReports),
                outputReports: reports(collection.outputReports),
                featureReports: reports(collection.featureReports)
            }))
        };
    }

    async function requestDevice(filters: any[]) {
        const api = hid();
        if (!api) return [];

        const requested = (filters ?? []).map((filter: any) => {
            const out: any = {};
            if (filter.vendorId != null) out.vendorId = filter.vendorId;
            if (filter.productId != null) out.productId = filter.productId;
            if (filter.usagePage != null) out.usagePage = filter.usagePage;
            if (filter.usage != null) out.usage = filter.usage;
            return out;
        }).filter((filter: any) => Object.keys(filter).length > 0);

        // requestDevice returns an array - the chooser can hand back more than one device when the
        // request allowed it - so this returns a list where the other buses return a single device.
        const devices = await api.requestDevice({ filters: requested });
        return devices.map((device: any) => describe(idOf(device), device));
    }

    async function getDevices() {
        const api = hid();
        if (!api) return [];
        try {
            const devices = await api.getDevices();
            return devices.map((device: any) => describe(idOf(device), device));
        } catch { return []; }
    }

    async function forget(id: string) {
        const device = _devices[id];
        if (!device || typeof device.forget !== 'function') return false;
        try { await device.forget(); release(id); return true; } catch { return false; }
    }

    function release(id: string) {
        const device = _registry.remove(id);
        if (!device) return;

        for (const key of Object.keys(_inputListeners)) {
            if (_inputListeners[key].device === device) unsubscribeInputReports(key);
        }
        // Fire-and-forget: disposal must not wait on a device that has already been unplugged.
        // close() rejects asynchronously, so the catch has to be on the promise as well - a bare
        // try/catch here would leave an unhandled rejection behind.
        try { if (device.opened) void device.close()?.catch(() => { /* already closed or gone */ }); }
        catch { /* already closed or gone */ }
    }

    async function open(id: string) {
        const device = _devices[id];
        if (!device) return false;
        if (device.opened) return true;
        await device.open();
        return true;
    }

    async function close(id: string) {
        const device = _devices[id];
        if (!device?.opened) return;
        try { await device.close(); } catch { /* already closed or gone */ }
    }

    function isOpened(id: string) { return !!_devices[id]?.opened; }

    function getInfo(id: string) {
        const device = _devices[id];
        return device ? describe(id, device) : null;
    }

    async function sendReport(id: string, reportId: number, data: Uint8Array) {
        const device = _devices[id];
        if (!device) return false;
        await device.sendReport(reportId, butil.utils.arrayToBuffer(data));
        return true;
    }

    async function sendFeatureReport(id: string, reportId: number, data: Uint8Array) {
        const device = _devices[id];
        if (!device) return false;
        await device.sendFeatureReport(reportId, butil.utils.arrayToBuffer(data));
        return true;
    }

    async function receiveFeatureReport(id: string, reportId: number) {
        const device = _devices[id];
        if (!device) return null;
        const view: DataView = await device.receiveFeatureReport(reportId);
        return butil.utils.viewToBytes(view);
    }

    function subscribeInputReports(subscriptionId: string, dotNetRef: any, id: string) {
        const device = _devices[id];
        if (!device) return false;

        const handler: EventListener = (event: any) => {
            const view: DataView = event.data;
            butil.utils.dispatch(dotNetRef, 'InvokeHidInputReport', subscriptionId, {
                reportId: event.reportId,
                data: butil.utils.viewToBytes(view)
            });
        };
        device.addEventListener('inputreport', handler);
        _inputListeners[subscriptionId] = { device, handler };
        return true;
    }

    function unsubscribeInputReports(subscriptionId: string) {
        const entry = _inputListeners[subscriptionId];
        if (!entry) return;
        delete _inputListeners[subscriptionId];
        entry.device.removeEventListener('inputreport', entry.handler);
    }

    // As in usb.ts: the event carries a device .NET may have no handle for, so one is minted here -
    // or the handle it already has is reused.
    function subscribeConnection(subscriptionId: string, dotNetRef: any) {
        const api = hid();
        if (!api) return false;

        const relay = (method: string) => ((event: any) => {
            butil.utils.dispatch(dotNetRef, method, subscriptionId, describe(idOf(event.device), event.device));
        }) as EventListener;

        const connect = relay('InvokeHidConnected');
        const disconnect = relay('InvokeHidDisconnected');
        api.addEventListener('connect', connect);
        api.addEventListener('disconnect', disconnect);
        _connectionListeners[subscriptionId] = { connect, disconnect };
        return true;
    }

    function unsubscribeConnection(subscriptionId: string) {
        const entry = _connectionListeners[subscriptionId];
        if (!entry) return;
        delete _connectionListeners[subscriptionId];
        const api = hid();
        if (!api) return;
        api.removeEventListener('connect', entry.connect);
        api.removeEventListener('disconnect', entry.disconnect);
    }
}(BitButil));
